### ISSUE TITLE
Only dependencies of one library target in Cabal project are loaded

### ISSUE THREAD

**Issue by liftM** [dramforever/vscode-ghc-simple#44]
## Status

- [x] A workaround is available
- [ ] A fix is written
- [ ] A fix is released
- [ ] The fixed version is on the marketplace.

---

- Version of vscode-ghc-simple: 0.1.13
- Version of VSCode: 1.40.0 (86405ea)
- Version of Haskell-related tools:
    - GHC: 8.6.5
    - Stack: N/A
    - Cabal: 3.0.0.0

## Steps to reproduce

I have a Haskell project `foo` that uses Cabal 3, not Stack. It has an executable component and a library component. In Cabal, these targets are `exe:foo` and `lib:foo`.

It looks like this extension correctly detects my project as Cabal, and shells out to `cabal repl`. `cabal repl` can only create a session for one target at a time, and defaults to the first library target of a project.

Targets `exe:foo` and `lib:foo` have different dependencies. For example, I keep my CLI parsing logic in `exe:foo`, so only `exe:foo` depends on `optparse-applicative`. However, since `cabal repl` by default opens a session for `lib:foo`, when I open a file in `exe:foo` in VS Code, I see this error:

```
Could not load module ‘Options.Applicative’
It is a member of the hidden package ‘optparse-applicative-0.15.1.0’.
Perhaps you need to add ‘optparse-applicative’ to the build-depends in your .cabal file.
Use -v to see a list of the files searched for.
```

This error is incorrect; `cabal build` still succeeds in this project. It occurs because `cabal repl` has loaded the dependencies for the `lib:foo` target, but VS Code is open to a file used in the `exe:foo` target. Since `cabal repl` has not loaded the dependencies unique to the `exe:foo` target, the session incorrectly believes that these modules are hidden.

This can be "worked around" by having all targets share all of their dependencies, but this is a pretty unacceptable workaround.

## Correct behaviour

The correct behaviour here would be to open unique GHCi sessions for each individual compilation target in the project, create a map of file names to compilation targets, and then pick the correct GHCi session for each particular file. In the rare case that a file is part of multiple compilation targets, any session should do.

#17 shares the same root cause as this bug, with the opposite effect. Since I'm using Cabal and `cabal repl` supports only one target, some of my files have incorrect missing dependency errors. In #17, the project uses Stack. When this extension is used with Stack projects, it tries to run `stack repl` on all targets listed by `stack ide targets`, which causes the naming conflicts seen in #17.

In both cases, the problem is that one GHCi session is being used for the whole project, when the correct behaviour would be to use one GHCi session per compilation target within the project.

**Comment by liftM** [comment]
Another side effect of this is that importing my library modules in my tests always cause a `Could not find module` error when viewing test files. This cannot be worked around: since my tests depend on my libraries, and the default `cabal repl` session opens on my library target, I can't add my tests as a dependency to my library because that would cause a circular dependency.

**Comment by dramforever** [comment]
Sorry for the extremely delayed response. I've recently started working on supporting multi-component/multi-package projects, i.e. the 'correct behavior' you have mentioned.

If you haven't abandoned vscode-ghc-simple, please consider testing out development versions of the `vgs-bios` branch. See more information in the discussion here: <https://github.com/dramforever/vscode-ghc-simple/discussions/87>.

In short, you will need to install a development version of the `vgs-bios` branch and set up an `hie.yaml` file, with the same format as the one accepted by [hie-bios] to tell vscode-ghc-simple about your project structure. If you need help or find any problems please do consider reaching out.

[hie-bios]: https://github.com/mpickering/hie-bios#explicit-configuration

Not everything is well-supported yet, but hopefully simple multi-component or multi-package setups will work. The same file should also help out those who use ghcide. Many projects already have an `hie.yaml` file, so these will hopefully work out of the box. See the discussion page linked above for details for more information.

Thanks for considering helping out!

**Comment by dramforever** [comment]
Released in v0.2.0
